File: leetcode/Maximum_Depth_of_Binary_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int max_depth;
    void findMaxDepth(TreeNode *node, int depth) {
        if (node == NULL)
            return;
        if (max_depth < depth)
            max_depth = depth;
        findMaxDepth(node->left, depth + 1);
        findMaxDepth(node->right, depth + 1);
    }
public:
    int maxDepth(TreeNode *root) {
        max_depth = 0;
        findMaxDepth(root, 1);
        return max_depth;
    }
};

File: leetcode/Maximum_Depth_of_Binary_Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    int max_height;
    void dfs(TreeNode t, int h) {
        if (t == null) return;
        if (h > max_height) max_height = h;
        dfs(t.left, h + 1);
        dfs(t.right, h + 1);
    }
    public int maxDepth(TreeNode root) {
        max_height = 0;
        dfs(root, 1);
        return max_height;
    }
}
